### DISCUSSION TITLE
[Ai-deck]Unable to run classification example

### DISCUSSION THREAD

**Discussion by zdh1995** [bitcraze/discussions#412] [Q&A]
Hi,
I try to run the classification exmaple by
` sudo docker run --rm -v ${PWD}:/module --privileged aideck-with-autotiler tools/build/make-example examples/ai/classification run`
and it stay at:
```
DEPRECATED! use 'adapter speed' not 'adapter_khz'
DEPRECATED! use 'adapter driver' not 'interface'
TARGET create
0
GAP8 INIT TARGET
GAP8 examine target
Init jtag
Initialising GAP8 JTAG TAP
Loading binary through JTAG
```
But when I try to flash it, it seems to be normal:

`sudo docker run --rm -v ${PWD}:/module --privileged aideck-with-autotiler tools/build/make-example examples/ai/classification flash`

```
DEPRECATED! use 'adapter speed' not 'adapter_khz'
DEPRECATED! use 'adapter driver' not 'interface'
TARGET create
GAP8 INIT TARGET
GAP8 examine target
Init jtag
Initialising GAP8 JTAG TAP
--------------------------
begining flash session (hyperflash)
--------------------------
load flasher to L2 memory
Loading binary through JTAG
[Flahser]: Hyperflash flasher is ready
Instruct flasher to begin flash per se
device struct address is 470047192
going to wait on addr GAP_RDY
wait on gap_rdy done witg buff ptr 0 469784844
loading image with addr 469784844 addr_min 469784844 and size 262144
load image done
wait on gap_rdy done witg buff ptr 0 469784844
loading image with addr 469522700 addr_min 469784844 and size 262144
load image done
wait on gap_rdy done witg buff ptr 0 469784844
loading image with addr 469260556 addr_min 469784844 and size 223504
load image done
[Flahser]: flasher is done
flasher is done, exiting
--------------------------
flasher is done!
--------------------------
--------------------------
Reset CONFREG to 0
--------------------------
GAP8 examine target
RESET: jtag boot mode=3
DEPRECATED! use 'adapter [de]assert' not 'jtag_reset'
Flashing image with command:
gap8-openocd -d0 -c "gdb_port disabled; telnet_port disabled; tcl_port disabled" -c "script interface/ftdi/olimex-arm-usb-tiny-h.cfg; script target/gap8revb.tcl; script tcl/flash_image.tcl; script tcl/jtag_boot.tcl; gap_flash_raw_hyper /module/examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img 747792 /gap_sdk/utils/openocd_tools; exit;"
```
I use the latest docker.
Any idea? 
Thank you~

**Comment by evoggy** [comment]
Hi,

When you use run the application it doesn't seem to work? If you want prinouts please uncomment this line (and comment the UART one): https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/ai/classification/Makefile#L14

**Reply by zdh1995** [reply]
Thank you for your reply!
I uncommnet the host line, and get some printout:

```
Opened Camera
Allocated memory for camera buffer
Allocated memory for output
Allocated memory for task
**Failed to construct CNN**
Launching execution with command:
gap8-openocd -d0 -c 'gdb_port disabled; telnet_port disabled; tcl_port disabled' -f interface/ftdi/olimex-arm-usb-tiny-h.cfg -f target/gap8revb.tcl -f tcl/jtag_boot_entry.tcl -c 'gap8_jtag_load_binary_and_start "/module/examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/classification" elf 0x1c000094'
Traceback (most recent call last):
  File "/gap_sdk/utils/gapy/gapy", line 162, in <module>
    main()
  File "/gap_sdk/utils/gapy/gapy", line 152, in main
    operationFunc(args, config, system)
  File "/gap_sdk/utils/gapy/run.py", line 184, in operationFunc
    raise RuntimeError('Runner has failed with value: %d' % status)
RuntimeError: Runner has failed with value: -1
make: *** [/gap_sdk/rtos/freeRTOS/vendors/gwt/rules/freeRTOS_rules.mk:413: run] Error 1
```
**It failed to construct CNN**. I print the return value of function classificationCNN_Construct(), and get error value **1**.
So, from the code:

```
	**AT_HYPERRAM_OPEN(&HyperRam, &HyperRamConf, &Error);
	if (Error) return 1;
	AT_HYPERFLASH_FS_OPEN(&HyperFlash, &HyperFlashConf, "classification_L3_Flash_Const.dat", &Error);
	if (Error) return 1;**
	classification_L3_Memory = (AT_HYPERRAM_POINTER) AT_HYPERRAM_ALLOC(&HyperRam, 304084);
	if (classification_L3_Memory == 0) return 2;
	classification_L2_Memory = (AT_L2_POINTER) AT_L2_ALLOC(0, 300000);
	if (classification_L2_Memory == 0) return 3;
	classification_L1_Memory = (AT_L1_POINTER) AT_L1_ALLOC(0, 46716);
	if (classification_L1_Memory == 0) return 4;
```
it seems there is something wrong in hyperRam or hyperFlash.

**Reply by zdh1995** [reply]
I further modified the function classificationCNN_Construct(),
```
	AT_HYPERRAM_OPEN(&HyperRam, &HyperRamConf, &Error);
	if (Error) return 1;
	AT_HYPERFLASH_FS_OPEN(&HyperFlash, &HyperFlashConf, "classification_L3_Flash_Const.dat", &Error);
	if (Error) return 2;
```
 and get the value, 2. So the error came from openning HyperFlash.
And I try to change the order of the error return to figure out if there were also some problems with  L1, L2, and L3 memory, but find that the program stopped at openning camera. However, when I test the test_camera example, it runs fine.

**Reply by evoggy** [reply]
Sorry for the late reply on this one. If I remember correctly you have to add an offset in the makefile because of our bootloader occupying the start of the flash. You also have to build using out docker container.

Try adding this to the makefile:
```
APP_CFLAGS += -DFS_PARTITIONTABLE_OFFSET=0x40000
```

**Reply by zdh1995** [reply]
Thank you for your reply!
And also sorry for the late reply, I tested positive for COVID-19.
I tried adding the code, and it didn't work.
And I search this code, it is already set in [https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/ai/classification/Makefile#L68](url)
What else can I try to figure out what's wrong?
Thank you!

**Comment by hmllr** [comment]
Hi :)
We looked into it and there were some issues introduced with the CPX framework which break this example. I just created a pull request https://github.com/bitcraze/aideck-gap8-examples/pull/107 which is fixing some bugs  -  the example should now work again and transmit you the classification output to the cfclient console (the printfs on gap8 are a bit unstable, so CPX is the more convenient way to get output from the example).
Hope this helps,
Hanna

**Reply by zdh1995** [reply]
Thank you for your reply and help！
I try to run the example with new code, but it still doesn't work.
In terminal, it stay at:
```
DEPRECATED! use 'adapter speed' not 'adapter_khz'
DEPRECATED! use 'adapter driver' not 'interface'
TARGET create
0
GAP8 INIT TARGET
GAP8 examine target
Init jtag
Initialising GAP8 JTAG TAP
Loading binary through JTAG
```
In the cfclient console,
```
SYS: Crazyflie 2.1 is up and running!
SYS: Production release 2020.04
SYS: I am 0x20303643305650090023002F and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_INFO: Warning! No driver found for deck.
DECK_CORE: 1 deck(s) found
IMU: BMI088 Gyro I2C connection [OK].
IMU: BMI088 Accel I2C connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Complementary (1) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
EEPROM: I2C connection [OK].
IMU: BMI088 gyro self-test [OK]
STAB: Wait for sensor calibration...
SYS: Free heap: 13216 bytes
STAB: Ready to fly.
```

**Reply by hmllr** [reply]
"DECK_INFO: Warning! No driver found for deck." seems to be the issue. Did you enable the AI-deck in the menuconfig? (you can check this by looking at your .config file in the build folder, it should have a line "CONFIG_DECK_AI=y")

**Comment by SUJUSTDO** [comment]
I have the same problem, is there a solution?

**Reply by Bananazooka** [reply]
Hi, could you elaborate a little about your problem? Is it exactly the same as explained above? If not, could you post a new discussion with as much information as possible? Thanks!

**Reply by SUJUSTDO** [reply]
Yes, I had the exact same problem.

**Reply by gemenerik** [reply]
Sorry to ask again, but could you clarify which specific problem you’re facing? This thread covers multiple issues, such as the deck driver not being found or the system getting stuck at initializing the GAP8, so it’s hard to tell exactly what you’re referring to. Additionally, since this discussion spans a long time and includes mentions of software updates, it’s important to know your current setup. What STM32 firmware version are you using? Are you running the latest version of the classification example? Have you tried other examples, and do they behave as expected? Providing these details will help us pinpoint the issue and suggest a solution.
